### PR TITLE
Add number of retries and error details for each step as a metric to telemetry

### PR DESCRIPTION
**Pull request by Viraj27**
# Description

This PR is for adding the following metrics to the telemetry payload:
1. No.of retries per each step of deployment.
2. Errors occurred for retries (if any) for each step of deployment.

## PR task checklist

- [x] I have performed a self-review of my code
- [x] I have followed code style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- ~[ ] I have made corresponding changes to the documentation~
- ~[] I have added thorough tests~
- [x] I have had a successful deployment with the changes in this PR

# Ticket

https://issues.redhat.com/browse/AAP-10571
https://issues.redhat.com/browse/AAP-10572

# Testing

1. Pull down the PR.
2. Create a new deployment using the installerincluded branch of aap-azurerm repo.
3. Once the Deployment is complete, go to Segment (https://app.segment.com/redhat-devtools/sources/ansible_on_azure_managed_app_dev/debugger) and check if the event is published here.
4. Check for the properties published in the PR.

# Sample Output

`client.Track(&analytics.Track{
  UserId: "3f7e29ba-24e0-42f6-8d9c-5149a14bda37",
  Event: "aap.azure.installer-deploy-success",
  Properties: map[string]interface{}{
    "accesstype": "public",
    "applicationid": "/subscriptions/3f7e29ba-24e0-42f6-8d9c-5149a14bda37/resourceGroups/vimehtaGroup/providers/Microsoft.Solutions/applications/AAPvimehta202304231323",
    "deploystatus": "succeeded",
    "endtime": "2023-04-23T21:35:40.000Z",
    "errors": map[string]interface{}{
      "AAP__Applications": "",
      "AAP__Operators": "",
      "AAP__Repository": "Code: Conflict - Message: {

  \"status\": \"Failed\",

  \"error\": {

    \"code\": \"ResourceDeploymentFailure\",

    \"message\": \"The 'AzureAsyncOperationWaiting' resource operation completed with terminal provisioning state 'Failed'.\",

    \"details\": [

      {

        \"code\": \"DeploymentFailed\",

        \"message\": \"At least one resource deployment operation failed. Please list deployment operations for details. Please see https://aka.ms/arm-deployment-operations for usage details.\",

        \"details\": [

          {

            \"code\": \"Conflict\",

            \"message\": \"{\r\n  \\"error\\": {\r\n    \\"code\\": \\"StorageAccountAlreadyExists\\",\r\n    \\"message\\": \\"The storage account named staapzzeq7nnljy7goeastus already exists under the subscription.\\"\r\n  }\r\n}\"

          }

        ]

      }

    ]

  }

}

Code: Conflict - Message: {

  \"status\": \"Failed\",

  \"error\": {

    \"code\": \"ResourceDeploymentFailure\",

    \"message\": \"The 'AzureAsyncOperationWaiting' resource operation completed with terminal provisioning state 'Failed'.\",

    \"details\": [

      {

        \"code\": \"DeploymentFailed\",

        \"message\": \"At least one resource deployment operation failed. Please list deployment operations for details. Please see https://aka.ms/arm-deployment-operations for usage details.\",

        \"details\": [

          {

            \"code\": \"Conflict\",

            \"message\": \"{\r\n  \\"error\\": {\r\n    \\"code\\": \\"StorageAccountAlreadyExists\\",\r\n    \\"message\\": \\"The storage account named staapzzeq7nnljy7goeastus already exists under the subscription.\\"\r\n  }\r\n}\"

          }

        ]

      }

    ]

  }

}

Code: Conflict - Message: {

  \"status\": \"Failed\",

  \"error\": {

    \"code\": \"ResourceDeploymentFailure\",

    \"message\": \"The 'AzureAsyncOperationWaiting' resource operation completed with terminal provisioning state 'Failed'.\",

    \"details\": [

      {

        \"code\": \"DeploymentFailed\",

        \"message\": \"At least one resource deployment operation failed. Please list deployment operations for details. Please see https://aka.ms/arm-deployment-operations for usage details.\",

        \"details\": [

          {

            \"code\": \"Conflict\",

            \"message\": \"{\r\n  \\"error\\": {\r\n    \\"code\\": \\"StorageAccountAlreadyExists\\",\r\n    \\"message\\": \\"The storage account named staapzzeq7nnljy7goeastus already exists under the subscription.\\"\r\n  }\r\n}\"

          }

        ]

      }

    ]

  }

}

Code: Conflict - Message: {

  \"status\": \"Failed\",

  \"error\": {

    \"code\": \"ResourceDeploymentFailure\",

    \"message\": \"The 'AzureAsyncOperationWaiting' resource operation completed with terminal provisioning state 'Failed'.\",

    \"details\": [

      {

        \"code\": \"DeploymentFailed\",

        \"message\": \"At least one resource deployment operation failed. Please list deployment operations for details. Please see https://aka.ms/arm-deployment-operations for usage details.\",

        \"details\": [

          {

            \"code\": \"Conflict\",

            \"message\": \"{\r\n  \\"error\\": {\r\n    \\"code\\": \\"StorageAccountAlreadyExists\\",\r\n    \\"message\\": \\"The storage account named staapzzeq7nnljy7goeastus already exists under the subscription.\\"\r\n  }\r\n}\"

          }

        ]

      }

    ]

  }

}
",
      "AKS__Cluster": "",
      "Application__Ingress": "",
      "Billing": "",
      "Database__Server__and__Databases": "",
      "Private__DNS": "",
      "Seeded__Content": "",
      "VNET__and__Subnets": "",
    },
    "region": "eastus",
    "retries": map[string]interface{}{
      "AAP__Applications": "0",
      "AAP__Operators": "0",
      "AAP__Repository": "4",
      "AKS__Cluster": "0",
      "Application__Ingress": "0",
      "Billing": "0",
      "Database__Server__and__Databases": "0",
      "Private__DNS": "0",
      "Seeded__Content": "0",
      "VNET__and__Subnets": "0",
    },
  },
})`

### CODE REVIEW

File: server/engine/telemetry.go [hunk 1]
```diff
@@ -27,11 +28,29 @@ func GetEvent(propertiesMap analytics.Properties) string {
...
+	retriesMap := make(map[string]interface{})
+	errorsMap := make(map[string]interface{})
 	var metricData []model.Telemetry
 	db.Find(&metricData)
+	// Error Details and Number of retries will be granular at the step level hence, each error and retry will be mapped to a step
+	// in a nested JSON like format
+	// Rest of the metrics are granular at the deployment level
 	for _, data := range metricData {
-		propertiesMap[string(data.MetricName)] = fmt.Sprintf("%v", data.MetricValue)
+		fmt.Printf("%+v\n", data)
```
Comment (tznamena) [inline server/engine/telemetry.go:39]: This seems like a left over from development.

File: server/engine/telemetry.go [hunk 2]
```diff
@@ -27,11 +28,29 @@ func GetEvent(propertiesMap analytics.Properties) string {
...
+		if data.Step == "" {
+			propertiesMap[string(data.MetricName)] = fmt.Sprintf("%v", data.MetricValue)
+		} else {
+			switch metric := data.MetricName; metric {
```
Comment (tznamena) [inline server/engine/telemetry.go:43]: This `switch` can be done without new variable:
```go
switch data.MetricName {
    ....
}
```

File: server/engine/telemetry.go [hunk 3]
```diff
@@ -27,11 +28,29 @@ func GetEvent(propertiesMap analytics.Properties) string {
...
 	var metricData []model.Telemetry
 	db.Find(&metricData)
+	// Error Details and Number of retries will be granular at the step level hence, each error and retry will be mapped to a step
+	// in a nested JSON like format
+	// Rest of the metrics are granular at the deployment level
 	for _, data := range metricData {
-		propertiesMap[string(data.MetricName)] = fmt.Sprintf("%v", data.MetricValue)
+		fmt.Printf("%+v\n", data)
+		if data.Step == "" {
+			propertiesMap[string(data.MetricName)] = fmt.Sprintf("%v", data.MetricValue)
```
Comment (tznamena) [inline server/engine/telemetry.go:41]: What's the need to call `fmt.Sprintf()` here? Can't we assign the metric value directly?

**Review by bhavenst — CHANGES_REQUESTED** [review]

File: server/engine/telemetry.go [hunk 4]
```diff
@@ -60,10 +123,11 @@ func PublishToSegment(db *gorm.DB) {
 		return
 	}
 	// set metrics in DB that are not set yet
-	model.SetMetric(db, model.ApplicationId, config.GetEnvironment().APPLICATION_ID)
+	model.SetMetric(db, model.ApplicationId, config.GetEnvironment().APPLICATION_ID, "")
 	// time.RFC3339 format is the Go equivalent to ISO 8601 format (minus the milliseconds)
-	model.SetMetric(db, model.EndTime, time.Now().Format(time.RFC3339))
+	model.SetMetric(db, model.EndTime, time.Now().Format(time.RFC3339), "")
```
[the remark this answers is not in the document]
Comment (Viraj27) [reply server/engine/telemetry.go:128]: Do you suggest renaming `PublishToSegment` to `BuildTelemetryAndPublishToSegment` ?

[the remark this answers is not in the document]
Comment (bhavenst) [reply server/engine/telemetry.go:128]: Two different ones, like SetFinalMetrics and PublishToSegment or something..

File: server/engine/telemetry.go [hunk 5]
```diff
@@ -41,17 +60,61 @@ func StoreMetricFromMainOutputs(db *gorm.DB) {
...
+}
+
+func StoreErrorsPerStep(db *gorm.DB, step model.Step) {
+
+	var errorsArr []string
+	var allErrors string
+
+	for _, execution := range step.Executions {
+		if execution.ErrorDetails != "" {
+			errorsArr = append(errorsArr, execution.ErrorDetails)
```
[the remark this answers is not in the document]
Comment (bhavenst) [reply server/engine/telemetry.go:91]: It looks like that is just a limitation of Metric (getter).  It works fine w/ the DB:

	db := persistence.NewInMemoryDB()
	model.SetMetric(db.Instance, model.Errors, "error1", "step1")
	model.SetMetric(db.Instance, model.Errors, "error2", "step1")

	retrieved := []model.Telemetry{}
	db.Instance.Find(&retrieved)
	for _, item := range retrieved {
		fmt.Printf("%v - %v - %v\n", item.Step, item.MetricName, item.MetricValue)
	}

=== RUN   TestTelemetryTable
step1 - errors - error1
step1 - errors - error2

File: server/engine/telemetry.go (hunk 1 above)
Comment (Viraj27) [reply server/engine/telemetry.go]: Removed this

File: server/engine/telemetry.go (hunk 2 above)
Comment (Viraj27) [reply server/engine/telemetry.go]: Updated

File: server/engine/telemetry.go (hunk 3 above)
Comment (Viraj27) [reply server/engine/telemetry.go]: You're right we can, when I was first trying this it was giving me some weird panic where it wasn't directly able to assign data.MetricValue directly to the map. But I tried it again now and it works.

File: server/engine/telemetry.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (Viraj27) [reply server/engine/telemetry.go]: Discussed offline and updated

File: server/engine/telemetry.go [hunk 6]
```diff
@@ -41,17 +60,61 @@ func StoreMetricFromMainOutputs(db *gorm.DB) {
...
+
+func StoreMetricsPerStep(db *gorm.DB) {
+
+	// Store number of retries and error details for each step if present
+	var allSteps []model.Step
+	db.Find(&allSteps)
+	db.Model(&model.Step{}).Preload("Executions").Find(&allSteps)
+
+	for _, step := range allSteps {
+		// len(step.Executions) includes the 1st attempt as well
```
[the remark this answers is not in the document]
Comment (Viraj27) [reply server/engine/telemetry.go]: Updated

File: server/engine/telemetry.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (Viraj27) [reply server/engine/telemetry.go:114]: Split it into two functions in the engine.go file

**Review by bhavenst — APPROVED** [review]

File: server/engine/telemetry.go [hunk 7]
```diff
@@ -41,29 +57,73 @@ func StoreMetricFromMainOutputs(db *gorm.DB) {
...
 	}
 	if access, exists := mainOutput.Values["access"]; exists {
-		model.SetMetric(db, model.AccessType, access.(map[string]interface{})["value"].(string))
+		model.SetMetric(db, model.AccessType, access.(map[string]interface{})["value"].(string), "")
 	} else {
 		log.Error("Access Type of deployment is missing : will not be included in telemetry")
 	}
 }
 
+func StoreRetriesPerStep(db *gorm.DB, step model.Step) {
```
Comment (tznamena) [inline server/engine/telemetry.go:71]: Does this function need to be exported(start with capital letter)? The only use is in `StoreMetricsPerStep()`.

File: server/engine/telemetry.go [hunk 8]
```diff
@@ -41,29 +57,73 @@ func StoreMetricFromMainOutputs(db *gorm.DB) {
...
+	// In cases where a step is not executed at all, retries should be set to 0.
+	retries := len(step.Executions) - 1
+	if retries < 0 {
+		retries = 0
+	}
+	model.SetMetric(db, model.Retries, fmt.Sprint(retries), step.Name)
+
+}
+
+func StoreErrorsPerStep(db *gorm.DB, step model.Step) {
```
Comment (tznamena) [inline server/engine/telemetry.go:84]: Same here, does this function need to be exported(start with capital letter)? The only use is in `StoreMetricsPerStep()`.

File: server/engine/telemetry.go [hunk 9]
```diff
@@ -41,29 +57,73 @@ func StoreMetricFromMainOutputs(db *gorm.DB) {
...
+
+	// for errors, loop through all executions and combine all errors found for each step and then merge them all
+	for _, execution := range step.Executions {
+		if execution.ErrorDetails != "" {
+			model.SetMetric(db, model.Errors, execution.ErrorDetails, step.Name)
+		}
+	}
+}
+
+func StoreMetricsPerStep(db *gorm.DB) {
```
Comment (tznamena) [inline server/engine/telemetry.go:94]: And here as well, does this function need to be exported(start with capital letter)? The only use is in `SetFinalMetrics()`.

Comment (Viraj27) [reply server/engine/telemetry.go]: Updated all function names and made them internal helpers

**Review by tznamena — APPROVED** [review]

**Pull request merged**